pkg/controller/workload/kubernetes/application/application.go

```diff
@@ -0,0 +1,342 @@
+	finalizerName    = "finalizer." + controllerName
+	reconcileTimeout = 1 * time.Minute
+
+	reasonGCResources     = "failed to garbage collect " + v1alpha1.KubernetesApplicationResourceKind
+	reasonSyncingResource = "failed to sync " + v1alpha1.KubernetesApplicationResourceKind
+)
+
+var log = logging.Logger.WithName("controller." + controllerName)
+
+type syncer interface {
+	sync(ctx context.Context, app *v1alpha1.KubernetesApplication) reconcile.Result
+}
+
+type deleter interface {
+	delete(ctx context.Context, app *v1alpha1.KubernetesApplication) reconcile.Result
+}
+
+// A syncdeleter can sync and delete resources in a KubernetesCluster.
+type syncdeleter interface {
+	syncer
+	deleter
+}
+
+// CreatePredicate accepts KubernetesApplications that have been scheduled to a
+// KubernetesCluster.
+func CreatePredicate(event event.CreateEvent) bool {
+	wl, ok := event.Object.(*v1alpha1.KubernetesApplication)
+	if !ok {
+		return false
+	}
+	return wl.Status.Cluster != nil
+}
+
+// UpdatePredicate accepts KubernetesApplications that have been scheduled to a
+// KubernetesCluster.
+func UpdatePredicate(event event.UpdateEvent) bool {
+	wl, ok := event.ObjectNew.(*v1alpha1.KubernetesApplication)
+	if !ok {
+		return false
+	}
+	return wl.Status.Cluster != nil
+}
+
+// Add the KubernetesApplication scheduler reconciler to the supplied manager.
+// Reconcilers are triggered when either the application or any of its resources
+// change.
+func Add(mgr manager.Manager) error {
+	kube := mgr.GetClient()
+	r := &Reconciler{
+		kube: kube,
+		local: &localCluster{
+			ar: &applicationResourceClient{kube: kube},
+			gc: &applicationResourceGarbageCollector{kube: kube},
+		},
+	}
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return errors.Wrap(err, "cannot create Kubernetes controller")
+	}
+
+	if err := c.Watch(
+		&source.Kind{Type: &v1alpha1.KubernetesApplicationResource{}},
+		&handler.EnqueueRequestForOwner{OwnerType: &v1alpha1.KubernetesApplication{}, IsController: true},
+	); err != nil {
+		return errors.Wrapf(err, "cannot watch for %s", v1alpha1.KubernetesApplicationResourceKind)
+	}
+
+	err = c.Watch(
+		&source.Kind{Type: &v1alpha1.KubernetesApplication{}},
+		&handler.EnqueueRequestForObject{},
+		&predicate.Funcs{CreateFunc: CreatePredicate, UpdateFunc: UpdatePredicate},
+	)
+	return errors.Wrapf(err, "cannot watch for %s", v1alpha1.KubernetesApplicationKind)
+}
+
+// localCluster is a syncDeleter that syncs and deletes resources from the same
+// cluster as their controlling application.
+type localCluster struct {
+	ar applicationResourceSyncer
+	gc garbageCollector
+}
+
+func (c *localCluster) sync(ctx context.Context, app *v1alpha1.KubernetesApplication) reconcile.Result {
+	// We want the garbage collector to delete all controlled resources when we
+	// delete the application.
+	util.AddFinalizer(app, metav1.FinalizerDeleteDependents)
```

Comment:
What is the use case for `foregroundDeletion`? I like the idea, just not sure if this is the intent.

Reply:
I think you're making the point here that my comment does not reflect what `foregroundDeletion` actually does, right?

I found the [garbage deletion docs](https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/) to be a little vague and had thought setting `foregroundDeletion` was required to ensure child resources would be deleted (rather than orphaned) by default. So my primary goal was to ensure child resources were deleted, not to block deletion of the parent _until_ the children were deleted.

https://github.com/kubernetes/kubernetes/blob/v1.14.1/pkg/controller/garbagecollector/garbagecollector.go#L502

Having skimmed the GC code just now it seems my understanding was mistaken and that the default is actually to background delete, rather than to orphan. I don't feel strongly about whether we foreground or background delete children. Let me know if you have an opinion and I'll either update the comment or remove the finalizer.

Reply:
Is this finalizer auto-removed, or do we have to remove it ourselves like the one below?

Reply:
It's removed by the Kubernetes garbage collector controller.

https://github.com/kubernetes/kubernetes/blob/v1.14.1/pkg/controller/garbagecollector/garbagecollector.go#L528

Reply:
I've removed this finalizer.